### PR TITLE
Fix Theano Cudnn BatchNorm when axis!=1

### PR DESCRIPTION
**Pull request by aloisg**
Adresses issues like : https://groups.google.com/forum/#!topic/keras-users/innS4NcLPt8

Seems like keras uses cudnn batch-norm "spatial" configuration, which normalizes over axis (0, 2, 3). When you don't use axis=1 in keras, theano tries to broadcast the axis dimension, and can't because this dimension is not 1.

One way to fix it is to shuffle dimensions to put the normalization axis over first dimension, theano batch-norm, and then shuffle it back.

I'm not sure this is the best way to do it, but I think it fixes the issue.

Results with different backends on https://github.com/keunwoochoi/music-auto_tagging-keras are now the same (they were not).

We should check the fix without cudnn too (but it should be no problem imo).

Also, it raises the question of default value of axis : for dense/conv2d, you'd want axis=1 for default (-1 would work for dense, not for conv2d), but for conv1d, you'd want axis=2 (or axis=-1).

### CODE REVIEW

**Comment by fchollet** [comment]
> Results with different backends on https://github.com/keunwoochoi/music-auto_tagging-keras are now the same (they were not).

When I've added support for this model in TF/Theano and with both dim orderings, I verified that outputs were always the same (in test mode).

It looks to me that this PR assumes that the Theano backend is being used. Can you comment on that?

**Comment by aloisg** [comment]
You're right, everything was working fine with tensorflow before this change, and we should move the fix in the theano backend (this fix is not changing the tensorflow result, but shouldn't be there anyway).

Theano bugged when using axis!=1, just like this : https://groups.google.com/forum/#!topic/keras-users/innS4NcLPt8. 
Keras uses the cudnn batch norm (http://deeplearning.net/software/theano_versions/dev/library/sandbox/cuda/dnn.html) with mode="spatial" which assumes axis=1, and then fails to broadcast (because axis 1 doesn't have dimension 1).

Experiments like https://github.com/keunwoochoi/music-auto_tagging-keras didn't fail though, cause the dimension on axis 1 was 1 imho. But the normalization was on the wrong axis. You can easily run the example with both backend on current keras version, and see the difference in results (which disappears with the fix).

**Comment by aloisg** [comment]
> When I've added support for this model in TF/Theano and with both dim orderings, I verified that outputs were always the same (in test mode).

Do you try it with cudnn at the time ?

**Comment by aloisg** [comment]
I'm not sure there is an issue with current keras version when we use http://deeplearning.net/software/theano/library/tensor/nnet/bn.html. Documentation about this isn't clear enough imho, I'm not really sure what it is doing when provided with parameters whose dim!=1 is not on the axis=1. I'll be doing more tests tomorrow, but problem could be related to the theano cudnn batch norm only.

**Comment by fchollet** [comment]
It looks like the issue is specific to cuDNN with the Theano backend. In that case it should be fixed only in the Theano backend, when we use cuDNN BN.

**Comment by aloisg** [comment]
I wonder if it is still useful to use cudnn batch_norm if we have to permute dimensions before and after (i.e when axis!=1) ? The performance gain may be lost ?

Also, as you suggested, it would be good to add the fix in the "batch_normalization" function of theano backend, but I'm not sure how to find which was the normalization axis. The easiest way would be to check for dim!=1 in mean shape for example, but I'm not really familiar with the way you can get shapes from theano tensor.

**Comment by aloisg** [comment]
I though looking for the non-broadcastable was a good solution, but seems the parameters happen to be broadcastable on all dimensions sometimes...

**Comment by fchollet** [comment]
> I though looking for the non-broadcastable was a good solution, but seems the parameters happen to be broadcastable on all dimensions sometimes...

We could avoid using cuDNN in that case. Do you know in which cases this can occur?

**Comment by aloisg** [comment]
It happened when the normalization axis had a dim of 1.

**Comment by aloisg** [comment]
Sorry for the delay here.
When none of the axis are broadcastable, use theano batch norm instead of cudnn.

**Comment by fchollet** [comment]
LGTM, thanks

**Comment by Dapid** [comment]
@aloisg thank you!
